### PR TITLE
RavenDB-16536 Making sure that we'll flush everything even if there is no activity on a database

### PR DESCRIPTION
**Pull request by arekpalinski**

### CODE REVIEW

File: src/Voron/Impl/Journal/WriteAheadJournal.cs [hunk 1]
```diff
@@ -576,6 +576,11 @@ public void OnTransactionCommitted(LowLevelTransaction tx)
             public long LastFlushedTransactionId => _lastFlushed.TransactionId;
             public long LastFlushedJournalId => _lastFlushed.JournalId;
             public long TotalWrittenButUnsyncedBytes => Interlocked.Read(ref _totalWrittenButUnsyncedBytes);
+
+            internal int TotalCommittedSinceLastFlushPages;
+            internal bool ShouldFlush => TotalCommittedSinceLastFlushPages != 0 || _lastFlushed.TransactionId != _waj._env.CurrentReadTransactionId; // TODO arek - double check the second condition
```
Comment (ayende) [inline src/Voron/Impl/Journal/WriteAheadJournal.cs]: Why the second part?

File: test/SlowTests/Voron/Issues/RavenDB_16536.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            Assert.True(Env.Journal.Applicator.ShouldFlush);
+
+            onLogsAppliedMre.Reset();
+
+
+            ownFlusher.ForTestingPurposesOnly().AddEnvironmentToFlushQueue(Env);
+
+            // below two lines are necessary
+            Thread.Sleep(1000);
+            Env.TimeToSyncAfterFlushInSec = 0;
```
Comment (ayende) [inline test/SlowTests/Voron/Issues/RavenDB_16536.cs]: Why is this?

File: test/SlowTests/Voron/Issues/RavenDB_16536.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+                    tx.Commit();
+                }
+            }
+
+            var onLogsAppliedMre = new ManualResetEventSlim(false);
+
+            Env.OnLogsApplied += () => onLogsAppliedMre.Set();
+
+            var ownFlusher = new GlobalFlushingBehavior();
```
Comment (ayende) [inline test/SlowTests/Voron/Issues/RavenDB_16536.cs:45]: why do you create an instance of this?

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: The problem was described here - https://issues.hibernatingrhinos.com/issue/RavenDB-16536

TLDR we had this integer - `SizeOfUnflushedTransactionsInJournalFile` (which I renamed to `TotalCommittedSinceLastFlushPages`) which was used to track if we need to flush or not but it was decreased even before we actually started the flush operation. It didn't take into account that we might not be able to release all committed pages because there might be read transactions running during the flush.

I'm a bit afraid of tracking _exactly_ how many of modified pages has been flushed (there could be some overwrites of modified pages so we're filtering them out etc) so decided that the easiest would be to check if we still have something to flush by tracking the last flushed transaction ID. As commented need to carefully verify that.

File: test/SlowTests/Voron/Issues/RavenDB_16536.cs (hunk 2 above)
Comment (arekpalinski) [reply test/SlowTests/Voron/Issues/RavenDB_16536.cs]: Sorry, forgot to finish the comments. We have internal checks which makes sure we don't flush too often. Anyway it should be set in the test during `Configure()` call

File: test/SlowTests/Voron/Issues/RavenDB_16536.cs (hunk 3 above)
Comment (arekpalinski) [reply test/SlowTests/Voron/Issues/RavenDB_16536.cs:45]: Because I want to manage and control it explicitly in the test

File: src/Voron/Impl/Journal/WriteAheadJournal.cs (hunk 1 above)
Comment (ayende) [reply src/Voron/Impl/Journal/WriteAheadJournal.cs]: okay, understood

File: test/SlowTests/Voron/Issues/RavenDB_16536.cs (hunk 2 above)
Comment (arekpalinski) [reply test/SlowTests/Voron/Issues/RavenDB_16536.cs]: No, we cannot set it via options because options set it to 30 seconds at minimum. Added the comment which explains why we have this.

**Review by ppekrol — APPROVED** [review]
